README.md

```diff
@@ -0,0 +1,237 @@
+
+
+
+## Goals
+
+* **Continue to provide an object abstraction for Docker containers**: This seems to have proven to be a popular way to interace with Docker from Java
+* **Allow Testcontainers to support test frameworks other than JUnit 4**: JUnit 5, Spock and TestNG would be the likely candidates
+* **Expose a reasonably simple API that emphasises discoverability of features and ease of use**: Again, the fluent setter pattern has proven to work well
+* **Allow us to clean up the Testcontainers API**: for example:
+  * making sure all fluent setters are consistent
+  * adding a separation between definition-time concerns and 'run-time' concerns
+  * add a plugin model to allow *composition* to be used in addition to *inheritance*
+* **Continue to support specialized container types via specialized classes** 
+* **(Secondary)** Publish the library as a general purpose API for interacting with containers, and make Testcontainers into a test-oriented layer on top. There are valid non-test scenarios for using this library. The two libraries would become:
+  * Container Core, a library focused on bridging the gaps between Docker and Java
+  * Testcontainers 2.x, a library focused on using Container Core for testing
+* **(Secondary, but important)** Even though there will be breaking API changes for Testcontainers 2.0, make best efforts to ensure a 1:1 mapping of concepts
+
+
+---
+
+##Runtime `Container` vs define-time `ContainerBuilder`
+
+Separating out the API for 'actions that are available when a container is being defined' and 'actions that are available when the container' is running is important. It will help reduce the API surface area to only the methods that make sense in a given context. This should improve ease of use and discoverability of the API, as well as resulting in a cleaner codebase.
+
+The `org.testcontainers.containercore` package shows the key classes for creating containers *that are not tied to any test framework*. Classes under this package would be **published as a ContainerCore JAR** (probably still under the org.testcontainers groupId, but a new artifact)
+
+### End result / usage
+
+The classes in this package allow for the following API and usage:
+
+```java
+// Building a simple container object not linked to any test framework
+final Container container = Container.builder()
+  .withImage("alpine:3.5")
+  .withExposedPort(1234)
+  .build();
+
+// now that the container is built, the `with___` methods disappear and runtime methods,
+//  like `exec` become available
+container.start();
+container.exec("date");
+container.stop();
+```
+
+
+
+### Implementation
+
+* `Container` holds configuration data in fields, plus methods for 'run-time' interactions such as `start`, `stop` , `exec`, etc.
+* `ContainerBuilder` (and its parent class `BaseContainerBuilder`)  exposes the define-time `withXYZ` fluent setter API
```

Comment:
What do you guys think about the immutable builder? With the current proposal builder is mutable and it could lead to some surprising behavior. For example, I want to have 2 PosgtreSQL containers who shares some common settings:
```java
final ContainerBuilder commonBuilder = PostgreSQLContainer.builder().withImage("posgtresql:9.6");
final Container pg1 = commonBuilder.withExposedPort(1234).build();
final Container pg2 = commonBuilder.withExposedPort(2345).build();
```
But this code would not work as expected because of mutable nature of builders.

I think immutability is always a good default for the libraries. In testcontainers we don't need high-performant builders, so extra allocations is not a problem.